api/authentication/patron_authentication_provider.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from flask import url_for
+from werkzeug.datastructures import Authorization
+
+from api.authentication.access_token import AccessTokenProvider
+from api.authentication.base import AuthenticationProvider
+from api.problem_details import PATRON_AUTH_ACCESS_TOKEN_INVALID
+from core.model import Patron, Session, get_one
+from core.util.problem_detail import ProblemDetail, ProblemError
+
+if TYPE_CHECKING:
+    from core.model import Library
+
+
+class PatronAccessTokenAuthenticationProvider(AuthenticationProvider):
+    """Patron Authentication based on a CM generated Access Token"""
+
+    FLOW_TYPE = "http://librarysimplified.org/authtype/basic-token"
+    DESCRIPTION = "Library Barcode + Token"
+
+    def __init__(self, _db: Session, library: Library):
+        self._db = _db
+        self.library_id = library.id
+        self.external_integration_id = AccessTokenProvider.get_integration(_db).id
+
+    def authenticated_patron(
+        self, _db: Session, token: dict | str
+    ) -> Patron | ProblemDetail | None:
+        """Authenticate the patron by decoding the JWE token and fetching the patron from the DB based on the patron ID"""
+
+        if type(token) is not str:
+            return None
+
+        try:
+            data = AccessTokenProvider.decode_token(_db, token)
+        except ProblemError as ex:
+            data = ex.problem_detail
+
+        if type(data) == ProblemDetail:
+            return data
+
+        # This exists because of mypy
+        assert type(data) is dict
+
+        try:
+            patron_id = data["id"]
+            # Ensure the password exists
+            data["pwd"]
+        except KeyError:
+            return PATRON_AUTH_ACCESS_TOKEN_INVALID
+
+        patron: Patron | None = get_one(_db, Patron, id=patron_id)
+        if patron is None:
+            return None
+
+        return patron
+
+    def get_credential_from_header(self, auth: Authorization) -> str | None:
+        """If we are the right type of token, then decode the password from the token"""
+        if (
+            auth
+            and auth.type.lower() == "bearer"
+            and AccessTokenProvider.is_access_token(auth.token)  # type: ignore[attr-defined]
+        ):
+            token = AccessTokenProvider.decode_token(self._db, auth.token)  # type: ignore[attr-defined]
+            if type(token) == dict:
+                return token.get("pwd")
+
+        return None
+
+    def _authentication_flow_document(self, _db):
+        """This auth type should not have an entry in the authentication document"""
+        token_url = url_for(
+            "patron_auth_token",
+            library_short_name=self.library(_db).short_name,
+            _external=True,
+        )
+        inputs = {"login": {"keyboard": "default"}, "password": {"keyboard": "default"}}
```

Comment:
Where would we set these values if we are not setting them through the UI? Or do we want to have this provider show up in the `patron_auth_services` ?

Reply:
Ok, I've changed the `settings` to inherit from the Basic Auth settings, and use those values in the authentication document. However I see no use of these values in the document since the access token provider should not have a UI for the values anyways, nor should it be presented to the user in any way.
It is a secondary authentication mechanism which does not have the features of the basic auth mechnism.

Reply:
Based on our slack conversation I have changed the Access token provider to shadow the basic auth provider, from inception to authentication document creation.